### ISSUE TITLE
Entity goal selector errors out when an entity dies due to a NPE, stopping my external plugin from listening to death events.

### ISSUE THREAD

**Issue by WyndevCodes** [SlimeDog/AggressiveAnimals#80]
**Stack Trace**
```
[17:56:14 ERROR]: Entity threw exception at dungeon-test:757932.6090062266,186.5,325.8306462977509
java.util.ConcurrentModificationException: null
        at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:756) ~[?:?]
        at java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:778) ~[?:?]
        at net.minecraft.world.entity.ai.goal.GoalSelector.tickRunningGoals(GoalSelector.java:152) ~[?:?]
        at net.minecraft.world.entity.ai.goal.GoalSelector.tick(GoalSelector.java:145) ~[?:?]
        at net.minecraft.world.entity.Mob.serverAiStep(Mob.java:920) ~[?:?]
        at net.minecraft.world.entity.LivingEntity.aiStep(LivingEntity.java:3351) ~[?:?]
        at net.minecraft.world.entity.Mob.aiStep(Mob.java:673) ~[?:?]
        at net.minecraft.world.entity.AgeableMob.aiStep(AgeableMob.java:155) ~[?:?]
        at net.minecraft.world.entity.animal.Animal.aiStep(Animal.java:63) ~[?:?]
        at net.minecraft.world.entity.animal.Wolf.aiStep(Wolf.java:186) ~[?:?]
        at net.minecraft.world.entity.LivingEntity.tick(LivingEntity.java:3006) ~[?:?]
        at net.minecraft.world.entity.Mob.tick(Mob.java:435) ~[?:?]
        at net.minecraft.world.entity.animal.Wolf.tick(Wolf.java:202) ~[?:?]
        at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1352) ~[?:?]
        at net.minecraft.world.level.Level.guardEntityTick(Level.java:1004) ~[?:?]
        at net.minecraft.server.level.ServerLevel.lambda$tick$12(ServerLevel.java:872) ~[?:?]
        at net.minecraft.world.level.entity.EntityTickList.forEach(EntityTickList.java:42) ~[paper-1.20.1.jar:git-Paper-196]
        at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:852) ~[?:?]
        at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1556) ~[paper-1.20.1.jar:git-Paper-196]
        at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:446) ~[paper-1.20.1.jar:git-Paper-196]
        at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1393) ~[paper-1.20.1.jar:git-Paper-196]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1170) ~[paper-1.20.1.jar:git-Paper-196]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:317) ~[paper-1.20.1.jar:git-Paper-196]
        at java.lang.Thread.run(Thread.java:840) ~[?:?]
```

**Reproduction steps**
It's pretty rare, but just kill a mob that has its AI modified by NMS. Sometimes it will give the above error, killing the mob but breaking all future bukkit listeners.

**Server Environment**
Server version: git-Paper-156 (MC: 1.20.1)
AggressiveAnimals version: 1.7.0

**Comment by SlimeDog** [comment]
We'll have a look, but we need more information. From the intake form:

**Server Environment**
 - OS [Linux, MacOS, Windows] version: 
- Java version [java -version]:
 - Server version [/version]: 
 - AggressiveAnimals version [/version AggressiveAnimals]: 

**Additional context**
 Link to your configuration:
 Link to error log:

**Comment by WyndevCodes** [comment]
**Server Environment**
OS version: I cannot find this on my hosting provider's webpage, but I'm guessing a recent Linux version.
Java version: Java 17
Server version: git-Paper-156 (MC: 1.20.1)
AggressiveAnimals version: 1.7.0

Configuration: [](https://pastebin.com/FEAQVcK7))
Error log: The error is shown above, and I can confirm it happens with mobs besides wolves. Any mob that is passive/neutral and has an AI modification can be subject to this issue.

**Comment by SlimeDog** [comment]
Thanks for the information. How often does it happen? It has not been experienced in our (extensive) testing, nor reported previously by other users.

**Comment by WyndevCodes** [comment]
It actually happens quite frequently. I myself have trouble reproducing it but I get players reporting it every hour or so, and sometimes I see the error in console every 10 minutes.

**Comment by SlimeDog** [comment]
Thanks. Is it the case that AggressiveAnimals entities are the only entities for which there are NBT modifications? And that the issue occurs when players kill otherwise passive/neutral entities that have attacked them via AggressiveAnimals? I would really like to see the configuration for the active mob-types, and the defaults.

I feel compelled to mention that 1.20.1 has more bugs generally than 1.20.2. Are you going to update?

**Comment by SlimeDog** [comment]
May we join your server?

**Comment by WyndevCodes** [comment]
To your first question, I'm certain this plugin is causing the issues. I ran it disabled for a small period of time and those who reported the dungeons failing reported no issues with the plugin disabled. There is no NMS modification on our end, and this plugin is the only external plugin we use for mob ai modification.
To your second question, yes. The only mobs that I get an error for are those mobs which are normally passive but are turned hostile.
Regarding the config, I thought I sent it above but GitHub seems to have removed the link. What a pain. I'll send it when I can later, but it is the default plugin config with always-aggressive set to true, baby aggression set to true, and around 7 mobs set to be aggressive.
Regarding updating the server, we may consider it if it will fix this issue. As of now we've had no reason to update.
And if you'd like to join the server, the IP is ```dungeondodge.net```.

**Comment by SlimeDog** [comment]
What does "dungeons failing" mean?

It would be helpful to know what other plugins are running. If you would rather not make that list public, please DM SlimeDog on Spigot.

**Comment by SlimeDog** [comment]
As a test, I just added `thorns 3` to my armor, created 100 chickens in an enclosed space (so they didn't wander away), and let them attack without me responding (except to heal myself occasionally). In pretty short order, they were all killed by thorn reflection as they attacked me. 

I ran this test five times, so 500 chickens. I did not encounter the ConcurrentModificationException, or any other errors. I think it safe to conclude that the issue is caused by the other plugins.
